src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs

```diff
@@ -0,0 +1,142 @@
+{
+    internal partial class CacheEntry
+    {
+        // this type exists just to reduce average CacheEntry size
+        // which typically is not using expiration tokens or callbacks
+        private sealed class CacheEntryTokens
+        {
+            private List<IChangeToken> _expirationTokens;
+            private List<IDisposable> _expirationTokenRegistrations;
+            private List<PostEvictionCallbackRegistration> _postEvictionCallbacks; // this is not really related to tokens, but was moved here to shrink typicall CacheEntry size
+
+            internal List<IChangeToken> ExpirationTokens => _expirationTokens ??= new List<IChangeToken>();
+            internal List<PostEvictionCallbackRegistration> PostEvictionCallbacks => _postEvictionCallbacks ??= new List<PostEvictionCallbackRegistration>();
+
+            internal void AttachTokens()
+            {
+                if (_expirationTokens != null)
+                {
+                    lock (this)
+                    {
+                        for (int i = 0; i < _expirationTokens.Count; i++)
+                        {
+                            IChangeToken expirationToken = _expirationTokens[i];
+                            if (expirationToken.ActiveChangeCallbacks)
+                            {
+                                if (_expirationTokenRegistrations == null)
+                                {
+                                    _expirationTokenRegistrations = new List<IDisposable>(1);
+                                }
+                                IDisposable registration = expirationToken.RegisterChangeCallback(ExpirationCallback, this);
+                                _expirationTokenRegistrations.Add(registration);
+                            }
+                        }
+                    }
+                }
+            }
+
+            internal bool CheckForExpiredTokens(CacheEntry cacheEntry)
+            {
+                if (_expirationTokens != null)
+                {
+                    for (int i = 0; i < _expirationTokens.Count; i++)
+                    {
+                        IChangeToken expiredToken = _expirationTokens[i];
+                        if (expiredToken.HasChanged)
+                        {
+                            cacheEntry.SetExpired(EvictionReason.TokenExpired);
+                            return true;
+                        }
+                    }
+                }
+                return false;
+            }
+
+            internal bool CanCopyTokens() => _expirationTokens != null;
+
+            internal void CopyTokens(CacheEntry parentEntry)
+            {
+                if (_expirationTokens != null)
+                {
+                    lock (this)
+                    {
+                        lock (parentEntry.GetOrCreateTokens())
+                        {
+                            foreach (IChangeToken expirationToken in _expirationTokens)
```

Comment:
Why in some cases are we using a for loop with direct indexing to enumerate the contents and in others we're using foreach?

Reply:
I don't know as I am not the author of the original code. To minimize the diff and make the review easier I've decided to change as few things as possible